---
title: "panCancerAsthma_mainAnalysis"
author: "Amrit Singh"
date: "December 19, 2016"
output: 
  html_document: 
    keep_md: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache = TRUE, warning = FALSE, message = FALSE)
```

# Import nanoString data, demographics and load libraries

```{r}
## logistics
root = "~/projects/asthma"; commandArgs <- function(...) root  # root directory, used for _dirs.R
source(paste0(root, "/code/_dirs.R"))
source(paste0(root, "/code/_func.R"))
source(paste0(root, "/code/_func-asthma.R"))
libr(append(pkgs(),c("ggrepel","mixOmics","gplots","igraph","nlme","cowplot","tidyverse")))


## load dataset
load(rnapc_datanew_dir)
```

# Demographics

```{r}
table(calculatedResponse = demoPre$calculated_Response, givenResponse = demoPre$givenResponse) ## all subjects had no replicates
table(calculatedResponse = demoPre$calculated_Response, Sex = demoPre$SEX)
table(calculatedResponse = demoPre$calculated_Response, Allergen = demoPre$Allergen_cleanLabel)
table(calculatedResponse = demoPre$calculated_Response, Site = demoPre$SITE)

demoPre$BMI <- demoPre$Wt..Kg./(demoPre$HT.cm./100)^2
demoPre$calculated_Response <- factor(demoPre$calculated_Response, levels = c("ER", "DR"))
variables  = c("AGE", "BMI", "BLFEV", "EAR", "LAR", "Leukocyte_Counts.x10.9.",
                "Neu_percent","lym_percent","mono_percent","eos_percent","baso_percent")
result <- descriptiveStat(demo=demoPre, groups = "calculated_Response", 
variables)
as.data.frame(result$meanSD)
as.data.frame(result$pval)

subset(demoPre[, c(variables, "calculated_Response")], calculated_Response == "DR")
```

# Principal Component Analysis

```{r, fig.width=8, fig.height=6}
variables  = c("Leukocyte_Counts.x10.9.","Neu_percent","lym_percent","mono_percent",
               "eos_percent","baso_percent", "AGE", "BMI", "EAR", "LAR", "SEX", 
               "Wt..Kg.", "HT.cm.", "BLFEV", "AIS")

compVar(demoPre, t(genEset), variables, ncomp = 20)$p + ggtitle("Genes")
```

# Lung Function
## Spirometry

```{r, fig.width=8, fig.height=5}
## FEV1 plot
demo.core.fev1 <- demoPre[, c("BLFEV", "F10L","F20L","F30L",
                                    "F45L","F60L", "F90L","F120L",
                                    "F180L","F240L","F300L","F360L","F420L")]
demo.core.fev1.scale <- 100*scale(t(demo.core.fev1), center=demo.core.fev1$BLFEV, scale=demo.core.fev1$BLFEV)

fev1.list <- list()
for(i in 1:ncol(demo.core.fev1.scale)){
  fev1.list[[i]] <- as.matrix(demo.core.fev1.scale[,i])
}
fev1.rbind <- do.call(rbind, fev1.list)

fev1.time <- as.data.frame(demo.core.fev1.scale)
fev1.time.melt <- gather(fev1.time)
fev1.dat <- data.frame(fev1=as.numeric(fev1.time.melt$value))
fev1.dat$Time <- rep(round(c(0,10,20,30,45,60,90,120,180,240,300,360,420)/60,2), ncol(demo.core.fev1.scale))
fev1.dat$Subj <- factor(rep(colnames(demo.core.fev1.scale), each=nrow(demo.core.fev1.scale)))
res <- as.character(demoPre$calculated_Response)
res[res == "ER"] <- "Early Responders"
res[res == "DR"] <- "Dual Responders"
fev1.dat$Response <- factor(rep(res, each=nrow(demo.core.fev1.scale)), levels=c("Early Responders","Dual Responders"))
col.subj <- factor(res, levels=c("Early Responders","Dual Responders"))
fev1.dat$Col <- factor(rep(col.subj, each=nrow(demo.core.fev1.scale)))

p1 <- ggplot(fev1.dat, aes(x = Time, y = fev1, group=Subj, colour=Col)) + geom_line() +
  stat_smooth(aes(group = 1)) + stat_summary(aes(group = 1),
                                             geom = "point", fun.y = mean, shape = 17, size = 3) + facet_grid(. ~ Response) +
  scale_y_continuous(expression('Percent drop in'~ FEV[1])) + #theme_bw() +
  scale_x_continuous(expression('Time (hours)')) +
  theme(axis.text.y = element_text(size = 15, hjust = 1)) + theme(axis.text.x = element_text(size = 15, hjust = 0.5))+
  theme(axis.title.x=element_text(size = 15)) + theme(axis.title.y=element_text(size = 15,angle = 90))+ 
  theme(plot.background = element_rect()) +   theme_bw() + 
  theme(strip.background = element_rect(colour = "black", fill = "white",
                                        size = 1), strip.text.x = element_text(size=20)) +
        theme(legend.position="none") + 
  geom_segment(aes(x = -0.08, y = -60, xend = -0.08, yend = -55), arrow = arrow(length = unit(0.3, "cm")), col = "black") +
  annotate("text", x=1.7, y=-58, label="Blood draw") +
  scale_color_manual(values=c("#00BFC4", "#F8766D"), name = "Response")
#pdf(paste0(rnapc_dir, "/results/Figures/Figure1A.pdf"), height = 5)
p1
#dev.off()

```

## AIS vs. LAR

```{r, fig.width=5, fig.height=5}
response <- demoPre$calculated_Response
col <- as.character(response)
col[col == "ER"] <- "#00BFC4"
col[col == "DR"] <- "#F8766D"
par(mar = c(5, 5, 3, 3))
lar <- demoPre$LAR
ais <- demoPre$AIS
ais[is.na(ais)] <- -1
#pdf(paste0(rnapc_dir, "/results/Figures/Figure1B.pdf"), height = 5, width = 5)
plot(lar ~ ais, col = col, 
     pch = as.numeric(factor(response))+14, axes = FALSE, ylim=c(-55,5), xlim=c(-1,8),
     xlab = expression("Allergen induced shift (pre PC"[20]~"/post PC"[20]~")"),
     ylab = expression("Maximum drop in FEV"[1]~"(3h-7h)"))
box()
axis(1, at=-1:8, labels=c("NA", paste(0:8)), col.axis="black", las=1)
axis(2, at=seq(-50, 10, length.out=7), labels=c(paste(seq(-50, 10, length.out=7))), col.axis="black", las=1)
rect(xleft=-2, ybottom=-10, xright=9, ytop=15, density = 10, angle = 45,
     col = "#00BFC4")
rect(xleft=-2, ybottom=-15, xright=2, ytop=-10, density = 10, angle = 45,
     col = "#00BFC4")
rect(xleft=-2, ybottom=-60, xright=9, ytop=-15, density = 10, angle = -45,
     col = "#F8766D")
rect(xleft=2, ybottom=-15, xright=9, ytop=-10, density = 10, angle = -45,
     col = "#F8766D")
text(x = 6, y = 3, labels = "ERs", cex = 3, col = "#00BFC4")
text(x = 6, y = -51, labels = "DRs", cex = 3, col = "#F8766D")
abline(h = -15, col = "black", lty = 2)
abline(v = 2, col = "black", lty = 2)
#dev.off()

```

# Differential gene expression

```{r, fig.width=7, fig.height=7}
design <- model.matrix(~response)
fit <- eBayes(lmFit(genEset, design))
top <- topTable(fit, coef = 2, adjust.method = "BH", n = nrow(fit))
top$logPval <- -log10(top$P.Value)
top$Gene <- rownames(top)
top$Class <- droplevels(ann[top$Gene, "Gene.Class"])

## Gail mentioned this gene (so I checked it out)
#plot(genEset["OSM",] ~ response,
#  ylab = expression("log"[2]~"counts"),
#  xlab = "Response", main = "OSM, p=0.13")

thres <- top$logPval[top$adj.P.Val == min(top$adj.P.Val)]
deGen <- ggplot(top, aes(x = logFC, y = logPval, color = Class)) + geom_point(size = 2) +
  geom_hline(aes(yintercept = thres[length(thres)]), lty = 2, col = "red") +
  geom_vline(aes(xintercept = 0), lty = 2, color = "gray") +
  geom_text_repel(data=filter(top, logPval >= thres[length(thres)]), aes(label=Gene)) + customTheme(sizeStripFont = 10, xAngle = 0, hjust = 0.5, vjust = 0.5, xSize = 15, ySize = 15, xAxisSize = 15, yAxisSize = 15) +
  scale_x_continuous(expression("log"[2]~"Fold-change"), limits = c(-0.8, 0.5)) +
  scale_y_continuous(expression("-log"[10]~"P-value")) +
  annotate("text", x = -0.5, y = 2.2, label = "BH-FDR=35.2%", col = "red") +
  ggtitle("Differential gene expression") + theme(legend.position = c(0.3, 0.9))
deGen
```

# Biomarker Analysis (all immune genes)

```{r}
nrow(ann); dim(genEset);
table(ann$Gene.Class)

## mRNA biomarker panel with optimal number of genes
id <- rownames(subset(ann, Gene.Class %in% c("CT Antigen", "Immune Response", "Immune Response - Checkpoint")))
X <- t(genEset[id,])
Y <- as.character(demoPre$calculated_Response)
Y[Y == "ER"] <- "earlyResponder"
Y[Y == "DR"] <- "dualResponder"
Y = factor(Y, levels = c("earlyResponder", "dualResponder"))
names(Y) <- rownames(X)
X.train <- X
Y.train <- Y
keepXgrid <- c(seq(10, 200, 15), 250, 300, 500)
ncomp <- 2
X.test = NULL; Y.test =  NULL
filter = "none"; topranked = 50
validation = "loo"; M =  2
iter = 2; threads = 2; progressBar = TRUE

#result <- tuned.sPLSDA(X.train, Y.train, keepXgrid, ncomp, X.test = X.test, Y.test = Y.test, 
#  filter = filter, topranked = topranked, validation = validation, M = M, 
#  iter = iter, threads = threads, progressBar = progressBar, optimal = "error", errorMethod = "max.dist")

#saveRDS(result, paste0(rnapc_dir, "/results/latePhaseBiomarkerPanel.rds"))
result <- readRDS(rnapc_panel_dir)
biomarkerPanel <- result$panel
#write.csv(biomarkerPanel, paste0(rnapc_dir, "/results/biomarkerPanelgenes.csv"))
selectedGenes <- genEset[biomarkerPanel, ]

## Determine classificatoin performance
#cv <- perf.tuned.sPLSDA(result)
#saveRDS(cv, paste0(rnapc_dir, "/results/panelPerformance.rds"))
cv <- readRDS(rnapc_panelperf_dir)
filter(cv$errorRate, Method == "max.dist", Index == "BER")
cv$perf
```

## Classification performances

```{r}
mrnaPerf.immRes <- lapply(keepXgrid, function(i){
  ncomp <- 2
  result.mrna <- splsda(X = X, Y = Y, keepX = rep(i, ncomp), ncomp = ncomp)
  cv <- as.data.frame(mixOmics::perf(result.mrna, validation = "loo", progressBar = FALSE)$error.rate)
  cv$ncomp <- paste("Comp", 1:ncomp, sep = ".")
  cv
})
#saveRDS(mrnaPerf.immRes, paste0(rnapc_dir, "/results/panelLOOCV.rds"))
mrnaPerf.immRes <- readRDS(rnapc_panelloocv_dir)

errGenesImmRes <- do.call(rbind, mrnaPerf.immRes)  %>% as.data.frame %>% 
  mutate(keepX = rep(keepXgrid, each = 2)) %>% 
  gather(Method, Error, -(ncomp:keepX)) %>% 
  filter(Method == "BER.max.dist") %>% 
  mutate(Dataset = "Transcriptomics") 
errGenesImmRes$ncomp[errGenesImmRes$ncomp == "Comp.2"] <- "Comp.1-2"

errGenesImmRes$Error[which.min(errGenesImmRes$Error)]
minErrKeepX <- errGenesImmRes$keepX[which.min(errGenesImmRes$Error)]
all(result$keepX == minErrKeepX)

#pdf(paste0(rnapc_dir, "/results/Figures/Figure2A.pdf"), height = 8, width = 8)
ggplot(errGenesImmRes, aes(x = keepX, y = Error, color = ncomp)) + geom_point() +
  geom_line() + facet_grid(.~Dataset, scales = "free") +
  geom_vline(aes(xintercept = minErrKeepX), lty = 2, color = "gray") +
  ggtitle("mRNA biomarker panel performance") +
  ylab("Leave-one-out cross-validation error rate") +
  xlab("Number of features in biomarker panel")
#dev.off()
```

## Sample plot

```{r}
result.mrna <- splsda(X = X, Y, keepX = result$keepX, ncomp = result$ncomp)

#pdf(paste0(rnapc_dir, "/results/Figures/Figure2B.pdf"), height = 6, width = 6)
plotIndiv(result.mrna, group = Y, ellipse = TRUE, star = TRUE,
  ind.names = FALSE, legend = TRUE, col.per.group = c("#00BFC4", "#F8766D"), 
  pch = 19, title = "mRNA biomarker panel", X.label = "Component 1", 
  Y.label = "Component 2")
#dev.off()
```

## heatmap of expression data of biomarker genes

```{r}
## heatmap
X <- scale(t(genEset[biomarkerPanel, ]))
X[which(X > 2)] <- 2
X[which(X < -2)] <- -2
col <- color.mixo(as.numeric(Y))
brewer.n = 8
brewer.name = "Dark2"
palette(brewer.pal(brewer.n, brewer.name))
Colcol=brewer.pal(brewer.n, brewer.name)[as.numeric(Y)]
colors = c(seq(-3,-2,length=100),seq(-2,0.5,length=100),seq(0.5,6,length=100))
my_palette <- colorRampPalette(c("green", "black", "red"))(n = 299)
#pdf(paste0(rnapc_dir, "/results/Figures/Figure3A.pdf"), height = 8, width = 8)
heatmap.2(X, trace="none",
          RowSideColors=Colcol,
          labRow = rep("", length(Colcol)),
           labCol = rep("", length(Colcol)),
          symm=F,symkey=F,symbreaks=T, scale="none", col=my_palette,
          margins=c(5,8), key.title = "Expression")
legend("topright", c("ER","DR"), col=unique(Colcol), pch=19)
#dev.off()
```

## Gene-Set enrichment analysis

```{r}
dfc <- dfAnn[dfAnn$GeneName %in% biomarkerPanel, ]
dfc$Pathway <- factor(as.character(dfc$Pathway), levels = names(table(dfc$Pathway)[order(table(dfc$Pathway))]))
dfc$GeneName <- factor(as.character(dfc$GeneName), levels = names(table(dfc$GeneName)[order(table(dfc$GeneName))]))

selectedPathway <- table(as.character(dfc$Pathway))[table(as.character(dfc$Pathway)) > 1]
dfc2 <- dfc[as.character(dfc$Pathway) %in% names(selectedPathway), ]

## up and down-regulated genes
fc0 <- rowMeans(selectedGenes[, Y == "dualResponder"])-rowMeans(selectedGenes[, Y == "earlyResponder"])
fc <- fc0
fc[fc0 > 0] <- "up"
fc[fc0 < 0] <- "down"
dfc2$FC <- factor(fc[as.character(dfc2$GeneName)], levels = c("up", "down"))

#pdf(paste0(rnapc_dir, "/results/Figures/Figure3B.pdf"), height = 8, width = 8)
ggplot(dfc2, aes(x = GeneName, y = Pathway, fill = FC)) + geom_tile() +
  geom_rect(aes(xmin = -Inf, 
                xmax = Inf, ymin = -Inf, ymax = Inf), fill = "black", 
                alpha = 0.03) +
  xlab("Biomarkers") + scale_fill_manual(values=c("red", "green")) +
  geom_tile() + 
  customTheme(sizeStripFont = 10, xAngle = 90, hjust = 1, vjust = 1, xSize = 0,
    ySize = 10, xAxisSize = 10, yAxisSize = 10) +
  theme(panel.background = element_rect(fill = "black")) +
  xlab("Biomarkers")
#dev.off()

## print to file (cd molecules and innate immune genes)
cdMol <- filter(dfc2, Pathway == "CD molecules")
nrow(cdMol)
#write.csv(cdMol, paste0(rnapc_dir, "/results/enrichr/biomarkerpanel/cdMoleculeslist.csv"))
## run enrichr on those list of genes (Human Gene Atlas results)
cdMol_humanGeneAtlas <- read.delim(rnapc_erhga_dir)
head(cdMol_humanGeneAtlas)

innImmuneGenes <- filter(dfc2, Pathway == "Innate immune response")
nrow(innImmuneGenes)
#write.csv(innImmuneGenes, paste0(rnapc_dir, "/results/enrichr/biomarkerpanel/innImmuneGenesList.csv"))
## run enrichr on those list of genes (Human Gene Atlas results)
innImmuneGenes_wikiPathways <- read.delim(rnapc_erwiki_dir)
head(innImmuneGenes_wikiPathways[innImmuneGenes_wikiPathways$Adjusted.P.value <0.05,])

```

# network analysis using biomarker panel genes and surrogate cell-type frequences based on SVD

```{r, fig.width=11, fig.height=11}
cellAnn <- subset(ann, !(Gene.Class %in% c("CT Antigen", "Immune Response", "Immune Response - Checkpoint")))
## cell-type modules
ccDat <- genEset[rownames(cellAnn), ] %>% as.data.frame %>% mutate(Cell = cellAnn$Cell.Type)
dim(ccDat)
ccDat$Cell <- droplevels(ccDat$Cell)

Cellmod0 <- lapply(levels(ccDat$Cell), function(i){
  prcomp(t(subset(ccDat, Cell == i)[, -ncol(ccDat)]), center = TRUE, scale. = TRUE)$x[,1,drop=FALSE]
})
ccPC <- do.call(cbind, Cellmod0)
colnames(ccPC) <- levels(ccDat$Cell)

cells <- factor(colnames(ccPC))
names(cells) <- colnames(ccPC)
ccGen <- factor(c(rep("Cells", ncol(ccPC)), rep("Genes", nrow(selectedGenes))))
names(ccGen) <- c(colnames(ccPC), rownames(selectedGenes))
## check order
all(colnames(selectedGenes) == rownames(ccPC))
## check overlapping genes
length(intersect(rownames(cellAnn), rownames(selectedGenes)))


dat <- cbind(t(selectedGenes), ccPC)
nperms = 1000
threads = 8
response <- as.character(demoPre$calculated_Response)
response[response == "ER"] <- -1
response[response == "DR"] <- 1
response <- as.numeric(response)
names(response) <- rownames(dat)

#t1 <- proc.time()
#dat <- dat[, apply(dat, 2, mad) > 0]
#fit.dingo <- fast.dingo(dat = dat, x = response, diff.score = TRUE, B = nperms, threads = threads)
#t2 <- proc.time()
#(t2-t1)/60  # over 10 hours!!
# user     system    elapsed 
#  2.846867   1.106917 613.670483 mintues
#saveRDS(fit.dingo, "~/Documents/Manuscript/panCancerAsthma_BigFilesNotOnGitHub/fit.dingoNanoString_1000perms.rds")
#dingoResult <- fit.dingo[c("genepair", "diff.score", "gene1", "gene2")]
#saveRDS(dingoResult, paste0(rnapc_dir, "/results/dingoResult.rds"))

cutOff = 2.5
fit.dingo <- readRDS(dingo_dir)
diff <- fit.dingo$genepair[abs(fit.dingo$diff.score)>cutOff,]
feat <- unique(c(as.character(diff$gene1), 
  as.character(diff$gene2)))

links <- rbind(diff)
colnames(links) <- c("from", "to")
links$weight <- fit.dingo$diff.score[abs(fit.dingo$diff.score)>cutOff]

shape <- rep("sphere", length(feat))
shape[feat %in% rownames(selectedGenes)] <- "circle"
size <- rep("15", length(feat))
size[feat %in% rownames(selectedGenes)] <- "10"
vertexSize <- rep("0.6", length(feat))
vertexSize[feat %in% rownames(selectedGenes)] <- "0.6"
nodes <- data.frame(id = feat,
                    x = rep(1, length(feat)),
                    shape = shape,
                    size = as.numeric(size),
                    color = c("#FFFFFF", "#9999CC")[as.numeric(factor(size, levels = c("15","10")))],
  vertexSize = as.numeric(vertexSize))

# Converting the data to an igraph object:
net <- graph.data.frame(links, nodes, directed = FALSE) 
colrs <- c("blue", "red")
corCutOff <- links$weight
corCutOff[corCutOff > 0] <- 1
corCutOff[corCutOff < 0] <- -1
E(net)$color <- colrs[as.numeric(as.factor(corCutOff))]

l <- layout.fruchterman.reingold(net)
l <- layout.norm(l, ymin=-1, ymax=1, xmin=-1, xmax=1)

cc <- table(as.character(dfAnn[dfAnn$Gene %in% feat, "Pathway"]))
mark.groups <- list()
for(i in 1 : length(cc)){
  mark.groups[[i]] <- which(feat %in% dfAnn$GeneName[dfAnn$Pathway == names(cc[i])])
}
names(mark.groups) <- names(cc)
mark.col <- rainbow(length(mark.groups), alpha=.5)
mark.groups2 <- mark.groups[unlist(lapply(mark.groups, length)) > 1 & unlist(lapply(mark.groups, length)) < 6]
mark.col2 <- mark.col[unlist(lapply(mark.groups, length)) > 1 & unlist(lapply(mark.groups, length)) < 6]

#pdf(paste0(rnapc_dir, "/results/Figures/Figure4.pdf"), height = 10, width = 10)
plot(net, rescale=F, layout=l*1.0,
  vertex.label.cex=nodes$vertexSize,
  vertex.label.color="black", 
  mark.groups=mark.groups2, mark.col=mark.col2, mark.border=NA,
  edge.curved=.1)
legend(x=0.5, y=-1, c("Cells", "Genes"), pch=c(21, 21), col = "black",
  pt.bg=c("white", "#9999CC"), pt.cex=2, cex=2, bty="n", ncol=1)
legend(x=-1.3, y=1.2, names(mark.groups2), pch=19,
  col=unlist(mark.col2), pt.bg=unlist(mark.col2), pt.cex=2, cex=1.3, bty="n", ncol=1)
legend(x=-1.3, y=-0.8, c("Conserved", "Differential"), pch=19,
  col=c("red", "blue"), pt.bg=c("red", "blue"), pt.cex=2, cex=2, bty="n", ncol=1,
  title = "Edge type")
#dev.off()

```

# Determine changes in gene expression between quiet vs. exacerbation and exacerbation vs. followup periods

```{r}
load(rnapc_geo_dir)
table(geoPanelGenes$phenotype)
# EXACERBATION    FOLLOW UP        QUIET 
#         118          100          118

```

## Changes in gene expression

```{r}
## quiet vs. exacerbations
geoEset_QuietvsExacer <- filter(geoPanelGenes, phenotype != "FOLLOW UP")
group <- factor(geoEset_QuietvsExacer$phenotype, levels = c("QUIET", "EXACERBATION"))
subject <- factor(geoEset_QuietvsExacer$donor)
genes_pval_QuietvsExacer <- list()
for(i in 3 : ncol(geoPanelGenes)){
  gene <- as.vector(as.matrix(geoEset_QuietvsExacer[, i]))
  df <- data.frame(gene, group, subject)
  groupedDat <- groupedData(gene ~ group | subject, data = df)
  fit <- lme(gene ~ group, data = groupedDat, random = ~ 1 | subject, na.action = na.omit)
  genes_pval_QuietvsExacer[[i-2]] <- coef(summary(fit))[2,]
}
names(genes_pval_QuietvsExacer) <- colnames(geoPanelGenes)[-c(1:2)]
genes_adjpval_QuietvsExacer <- as.data.frame(do.call(rbind, genes_pval_QuietvsExacer))
genes_adjpval_QuietvsExacer$BH.FDR <- p.adjust(genes_adjpval_QuietvsExacer$`p-value`, "BH")
genes_adjpval_QuietvsExacer <- genes_adjpval_QuietvsExacer[order(genes_adjpval_QuietvsExacer$`p-value`), ]

genes_genSym_QuietvsExacer <- rownames(genes_adjpval_QuietvsExacer)[genes_adjpval_QuietvsExacer$BH.FDR < 0.1]
length(genes_genSym_QuietvsExacer)

genes_quiet.exacerUp <- rownames(genes_adjpval_QuietvsExacer)[genes_adjpval_QuietvsExacer$Value > 0 & genes_adjpval_QuietvsExacer$BH.FDR < 0.1]
genes_quiet.exacerDown <- rownames(genes_adjpval_QuietvsExacer)[genes_adjpval_QuietvsExacer$Value < 0 & genes_adjpval_QuietvsExacer$BH.FDR < 0.1]
length(genes_quiet.exacerUp); length(genes_quiet.exacerDown);


# exacerbation vs. followup
geoEset_ExacerVsFollowUp <- filter(geoPanelGenes, phenotype != "QUIET")
group <- factor(geoEset_ExacerVsFollowUp$phenotype, levels = c("EXACERBATION", "FOLLOW UP"))
subject <- factor(geoEset_ExacerVsFollowUp$donor)
genes_pval_ExacerVsFollowUp <- list()
for(i in 3 : ncol(geoPanelGenes)){
  gene <- as.vector(as.matrix(geoEset_ExacerVsFollowUp[, i]))
  df <- data.frame(gene, group, subject)
  groupedDat <- groupedData(gene ~ group | subject, data = df)
  fit <- lme(gene ~ group, data = groupedDat, random = ~ 1 | subject, na.action = na.omit)
  genes_pval_ExacerVsFollowUp[[i-2]] <- coef(summary(fit))[2,]
}
names(genes_pval_ExacerVsFollowUp) <- colnames(geoPanelGenes)[-c(1:2)]
genes_adjpval_ExacerVsFollowUp <- as.data.frame(do.call(rbind, genes_pval_ExacerVsFollowUp))
genes_adjpval_ExacerVsFollowUp$BH.FDR <- p.adjust(genes_adjpval_ExacerVsFollowUp$`p-value`, "BH")
genes_adjpval_ExacerVsFollowUp <- genes_adjpval_ExacerVsFollowUp[order(genes_adjpval_ExacerVsFollowUp$`p-value`), ]

genes_genSym_ExacerVsFollowUp <- rownames(genes_adjpval_ExacerVsFollowUp)[genes_adjpval_ExacerVsFollowUp$BH.FDR < 0.1]
length(genes_genSym_ExacerVsFollowUp)

genes_exacer.followUp <- rownames(genes_adjpval_ExacerVsFollowUp)[genes_adjpval_ExacerVsFollowUp$Value > 0 & genes_adjpval_ExacerVsFollowUp$BH.FDR < 0.1]
genes_exacer.followDown <- rownames(genes_adjpval_ExacerVsFollowUp)[genes_adjpval_ExacerVsFollowUp$Value < 0 & genes_adjpval_ExacerVsFollowUp$BH.FDR < 0.1]
length(genes_exacer.followUp); length(genes_exacer.followDown);

```

## plot significant genes

```{r, fig.width=14}
int <- intersect(genes_quiet.exacerUp, genes_exacer.followDown)
length(int)
#write.csv(int, paste0(rnapc_dir, "/results/enrichr/asthmaExacerbationGenes/asthmaExacerbationGenes.csv"))

## Quiet vs. Excerbation
genes_adjpval_QuietvsExacer$logPval <- -log10(genes_adjpval_QuietvsExacer$`p-value`)
genes_adjpval_QuietvsExacer$Gene <- rownames(genes_adjpval_QuietvsExacer)
commGene <- rep("common", nrow(genes_adjpval_QuietvsExacer))
commGene[!(genes_adjpval_QuietvsExacer$Gene %in% int)] <- "distinct"
genes_adjpval_QuietvsExacer$Overlap <- commGene
## volcano plot
thres <- genes_adjpval_QuietvsExacer$logPval[genes_adjpval_QuietvsExacer$BH.FDR < 0.1]
gg1 <- ggplot(genes_adjpval_QuietvsExacer, aes(x = Value, y = logPval, color = Overlap)) + geom_point(size = 2) +
  geom_hline(aes(yintercept = thres[length(thres)]), lty = 2, col = "gray") +
  geom_vline(aes(xintercept = 0), lty = 2, color = "gray") +
  geom_text_repel(data=filter(genes_adjpval_QuietvsExacer, logPval >= thres[length(thres)]), aes(label=Gene)) + customTheme(sizeStripFont = 10, xAngle = 0, hjust = 0.5, vjust = 0.5, xSize = 15, ySize = 15, xAxisSize = 15, yAxisSize = 15) +
  scale_x_continuous(expression("log"[2]~"fold-change (Exacer - Quiet)")) +
  scale_y_continuous(expression("-log"[10]~"P-value")) +
  annotate("text", x = 0.7, y = 1.7, label = "BH-FDR=10%", col = "red") +
  ggtitle("Quiet vs. Exacerbation") + theme(legend.position = "none") +
  scale_colour_brewer(palette="Set1")


## Exacerbation vs. Follow-up
genes_adjpval_ExacerVsFollowUp$logPval <- -log10(genes_adjpval_ExacerVsFollowUp$`p-value`)
genes_adjpval_ExacerVsFollowUp$Gene <- rownames(genes_adjpval_ExacerVsFollowUp)
commGene <- rep("common", nrow(genes_adjpval_ExacerVsFollowUp))
commGene[!(genes_adjpval_ExacerVsFollowUp$Gene %in% int)] <- "distinct"
genes_adjpval_ExacerVsFollowUp$Overlap <- commGene
## volcano plot
thres2 <- genes_adjpval_ExacerVsFollowUp$logPval[genes_adjpval_ExacerVsFollowUp$BH.FDR < 0.1]
gg2 <- ggplot(genes_adjpval_ExacerVsFollowUp, aes(x = Value, y = logPval, color = Overlap)) + geom_point(size = 2) +
  geom_hline(aes(yintercept = thres2[length(thres2)]), lty = 2, col = "gray") +
  geom_vline(aes(xintercept = 0), lty = 2, color = "gray") +
  geom_text_repel(data=filter(genes_adjpval_ExacerVsFollowUp, logPval >= thres2[length(thres2)]), aes(label=Gene)) + customTheme(sizeStripFont = 10, xAngle = 0, hjust = 0.5, vjust = 0.5, xSize = 15, ySize = 15, xAxisSize = 15, yAxisSize = 15) +
  scale_x_continuous(expression("log"[2]~"fold-change (Followup - Exacer)")) +
  scale_y_continuous(expression("-log"[10]~"P-value")) +
  annotate("text", x = -0.7, y = 1.8, label = "BH-FDR=10%", col = "red") +
  ggtitle("Exacerbation vs. Follow-up") + 
  theme(legend.position = c(0.85, 0.9)) +
  scale_colour_brewer(palette="Set1")

#pdf(paste0(rnapc_dir, "/results/Figures/Figure5.pdf"), height = 5, width = 11)
plot_grid(gg1, gg2, labels = c("A", "B"))
#dev.off()

```

## Gene set enrichment of the 28 asthma exacerbation genes

```{r}
genesetEnrich <- read.delim(rnapc_erreact_dir)
genesetEnrich[genesetEnrich$Adjusted.P.value < 0.05, ]

```


## Lincs1000 Chemical Perturbations

```{r}
lincsChemDown <- read.delim(rnapc_erexach_dir)
topDrugs <- lincsChemDown[lincsChemDown$Adjusted.P.value < 0.05, -c(5:6)]
topDrugs$Overlap <- gsub("/", " out of ", topDrugs$Overlap)
topDrugs$P.value <- format(topDrugs$P.value, digits=1)
topDrugs$Adjusted.P.value <- format(topDrugs$Adjusted.P.value, digits=1)
#write.csv(topDrugs, "~/Documents/Manuscript/geneMet/ARJCCM/figs/topDrugs.csv")

drugDat <- as.data.frame(do.call(rbind, apply(topDrugs, 1, function(i){
  genes <- strsplit(as.character(as.matrix(i))[ncol(topDrugs)], ";")[[1]]
  drugs <- rep(as.character(as.matrix(i))[1], length(genes))
  cbind(drugs, genes)
})))
drugDat$drugs <- unlist(lapply(strsplit(as.character(drugDat$drugs), "-"), function(i){
  paste(i[-c(1, length(i))], collapse = "-")
}))
drugDat$drugs <- factor(drugDat$drugs, levels = names(table(drugDat$drugs)[order(table(drugDat$drugs))]))
drugDat$genes <- factor(drugDat$genes, levels = names(table(drugDat$genes)[order(table(drugDat$genes))]))
#write.csv(names(table(drugDat$drugs)[order(table(drugDat$drugs), decreasing = TRUE)]), "~/Documents/Manuscript/geneMet/ARJCCM/figs/LINCSL1000_drugs_cdrd.csv")

## make heatamp
lincs1 <- ggplot(drugDat, aes(x = drugs, y = genes)) + geom_tile() + customTheme(sizeStripFont = 10, xAngle = 30, hjust = 1, vjust = 1, xSize = 10, ySize = 10, xAxisSize = 15, yAxisSize = 15) +
  xlab("LINCS L1000 Chem. Pert. down") + ylab("Genes")
```

## Lincs1000 Ligand Perturbations up

```{r}
lincsLigandUp <- read.delim(rnapc_erexali_dir)
lincsLigandUp.significant <- lincsLigandUp[lincsLigandUp$Adjusted.P.value < 0.05, ]

ligandDat <- as.data.frame(do.call(rbind, apply(lincsLigandUp.significant, 1, function(i){
  genes <- strsplit(as.character(as.matrix(i))[ncol(lincsLigandUp)], ";")[[1]]
  ligand <- rep(as.character(as.matrix(i))[1], length(genes))
  cbind(ligand, genes)
}))) %>% 
  separate(., col = "ligand", into = c("ligand", "CellCulture"), sep = "-")
ligandDat$ligand <- factor(ligandDat$ligand, levels = names(table(ligandDat$ligand)[order(table(ligandDat$ligand))]))
ligandDat$genes <- factor(ligandDat$genes, levels = names(table(ligandDat$genes)[order(table(ligandDat$genes))]))

lincs2 <- ggplot(ligandDat, aes(x = ligand, y = genes)) + geom_tile() + customTheme(sizeStripFont = 10, 
  xAngle = 0, hjust = 0.5, vjust = 0.5, xSize = 10, ySize = 10, xAxisSize = 15, yAxisSize = 15) +
  xlab("LINCS L1000 Ligand Pert. up") + ylab("Genes")


#pdf(paste0(rnapc_dir, "/results/Figures/Figure6.pdf"), height = 7, width = 11)
plot_grid(lincs1, lincs2, labels = c("A", "B"))
#dev.off()
```

