## traceR assess the enrichment level of taSNPs in a given sets of genomic intervals
library("traseR")


### 1. prep genomics interval with columns: chr, start, end (or GRanges object)
data(Tcell)




### 2. obtain taSNPs (trait associated)

## traceR specifies taSNPs objects with columns:
# 1. Trait: Description of disease/trait examined in the study
# 2. Trait_Class: Trait class which is formed based on the phenotype tree. Close traits are grouped together to form one class.
# 3. SNP_ID: SNP rs number
# 4. p.value: GWAS reported p-values
# 5. seqnames: Chromosome number associated with rs number
# 6. ranges: Chromosomal position, in base pairs, associated with rs number
# 7. Context: SNP functional class
# 8. GENE_NAME: Genes reported to be associated with SNPs
# 9. GENE_START: Chromosome start position of genes
# 10. GENE_END: Chromosome end position of genes
# 11. GENE_STRAND: Chromosome strand associated with SNPs


## traceR auto synchs taSNP from Association Results Browser (http://www.ncbi.nlm.nih.gov/projects/gapplusprev/sgap_plus.htm) 
# combines identified taSNPs from dbGaP and NHGRI, which together provide 
# 44,078 SNP-trait associations, 
# 48,936 SNP-trait class associations, 
# 30,553 unique taSNPs, 
# 573 unique traits and 
# 33 unique trait classes
data(taSNP)

## CEU vcf files (all seq var info) from (ftp://share.sph.umich.edu/1000genomes/fullProject/2012.03.14/)
# id linkage disequi SNPs>.8 located within 100kb of taSNP
# 1. vcftools: convert vcf file to PLINK
# 2. PLINK: call LD taSNPs by limitting linkage diseq SNPs >.8 (-ld-window-r2 0.8) within 100 kb of taSNP (-ld-window-kn 100)
# result: we have 90,700 SNP-trait associations and 78,247 unique linkage disequilibrium trait-associated SNP
system("vcftools -vcf vcf.file -out plink.file -plink plink -file plink.file -r2 -inter-chr -ld-snp-list snps.txt -ld-window-r20.8 -ld-window-kb 100 -out output.file-noweb")

# We also build linkage disequilibrium taSNP into another GRanges object taSNPLD, it has columns:
# 1. seqnames: Chromosome number associated with rs number
# 2. SNP_ID: SNP rs number
# 3. ranges: Chromosomal position, in base pairs, associated with rs number
# 4. Trait: Description of disease/trait examined in the study
# 5. Trait_Class: Trait class which is formed based on the phenotype tree. Close traits are grouped together to form one class.
data(taSNPLD)

## 1000 genome project
# get all SNPs with MAF (minor allele freq) from CEU vcf files downloaded
# totals 6,571,512 SNPs (MAF>0.05) excluding variants on Y chromosome = background in hypothesis testing
# we build these SNPs into GRanges object CEU:
# 1. seqnames: Chromosome number associated with rs number
# 2. SNP_ID: SNP rs number
# 3. ranges: Chromosomal position, in base pairs, associated with rs number
system("plink -file plink.file -freq -out chr")
data(CEU)




### 4. using traceR ro query a set of a genomic intervals against all taSNPs; explore genes/SNPs of interest

# whole genome: asume each base can be a taSNP; 
#  can classify all bases as is taSNP or not & in genomic interval or not
# all SNP: assume each SNP can be a taSNP; 
#  can classify all SNPs as is taSNP or not & in genomic interval or not

# can either 
#  in/exclude LD SNPs as taSNPs & 
#  use whole genome or all SNP as background for hypothesis testing

# region: data frame
# snpdb: taSNPs or including LD SNPs
# snpdb.bg: background SNP
# rankby (e.g. "pvalue", "odds.ratio"): sort by
# test.method (e.g. 
#  "binomial" (default: varying sample size/distr assumption/time): assume %observe a single SNP/base being a taSNP = in(#taSNP/#allSNP)&out of genomic interval, 
#  "chisq": (#s of contingency tableare large and balanced) difference of taSNP in & out of genomic intervals; assume those in genomic intervals follow hypergeometric distribution & use fisher's exact test to get p value, 
#   "fisher": (#s of contingency table is small < 20) , 
#   "nonparametric": (# of query is small < 1000): matched genomic intervals are generated by permuting genomic intervals randomly N times overlapping with taSNPs each time; then we calc empirical p value by counting how many taSNP hits >/< observed taSNP hits)

# alternative (e.g. "greater": test whether genomic intervals are enriched of taSNPs than bg, 
#  "less": test whether genomic intervals are depleted of taSNPs than bg)
x = traseR(snpdb=taSNP,  region=Tcell) # whole gnome
print(x)

x = traseR(snpdb=taSNPLD,region=Tcell) # include LD SNP as background
print(x)

x = traseR(snpdb=taSNP,  region=Tcell, snpdb.bg=CEU) # use all SNPs as background






### 5. results

# plot distribution of SNP functional class
plotContext(snpdb=taSNP, region=Tcell, keyword="Autoimmune")

# plot distribution of p value of taSNPs
plotPvalue(snpdb=taSNP, region=Tcell, keyword="autoimmune", plot.type="densityplot")

# plot SNP/genes given genomiv interval
plotInterval(snpdb=taSNP,data.frame(chr="chrX",start=152633780,end=152737085))

# query trait-associated SNPs by key word
x=queryKeyword(snpdb=taSNP,region=Tcell,keyword="autoimmune",returnby="SNP")
head(x)

# query trait-associated SNPs by gene name
x=queryGene(snpdb=taSNP,genes=c("AGRN","UBE2J2","SSU72"))

# query trait-associated SNPs by SNP name
x=querySNP(snpdb=taSNP,snpid=c("rs3766178","rs880051"))
x




